From 9935d2fccb4206c2949b43cc118aad6789303f12 Mon Sep 17 00:00:00 2001
From: shbatm <tim@shbatm.com>
Date: Fri, 13 Oct 2017 13:29:39 -0500
Subject: [PATCH] Gracefully shutdown node_helpers

---
 js/app.js      | 20 ++++++++++++++++++++
 js/electron.js | 10 ++++++++++
 2 files changed, 30 insertions(+)

diff --git a/js/app.js b/js/app.js
index cf11e4b..9042fe4 100644
--- a/js/app.js
+++ b/js/app.js
@@ -236,6 +236,26 @@ var App = function() {
 			});
 		});
 	};
+
+	/* stop()
+	 * This methods stops the core app.
+	 * This calls each node_helper's STOP function, if it exists.
+	 */
+	this.stop = function() {
+		for (var h in nodeHelpers) {
+			var nodeHelper = nodeHelpers[h];
+			if (typeof nodeHelper.stop === "function") {
+				nodeHelper.stop();
+			}
+		}
+	};
+
+	process.on("SIGINT", () => {
+		console.log("[SIGINT] Received. Shutting down server...");
+		setTimeout(() => { process.exit(0); }, 3000);  // Force quit after 3 seconds
+		this.stop();
+		process.exit(0);
+	});
 };
 
 module.exports = new App();
diff --git a/js/electron.js b/js/electron.js
index 84842ed..6d3713a 100644
--- a/js/electron.js
+++ b/js/electron.js
@@ -96,6 +96,16 @@ app.on("activate", function() {
 	}
 });
 
+// This method will be called when SIGINT is received and will call
+// each node_helper's stop function if it exists.
+app.on("before-quit", (event) => {
+  console.log("Shutting down server...");
+  event.preventDefault();
+  setTimeout(() => { process.exit(0); }, 3000); // Force-quit after 3 seconds.
+  core.stop();
+  process.exit(0);
+});
+
 // Start the core application if server is run on localhost
 // This starts all node helpers and starts the webserver.
 if (["localhost", "127.0.0.1", "::1", "::ffff:127.0.0.1", undefined].indexOf(config.address) > -1) {
-- 
2.1.4

